compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs

```diff
@@ -40,17 +43,29 @@ struct ConstToPat<'a, 'tcx> {
     span: Span,
     param_env: ty::ParamEnv<'tcx>,
 
-    // This tracks if we signal some hard error for a given const value, so that
+    // This tracks if we emitted some hard error for a given const value, so that
     // we will not subsequently issue an irrelevant lint for the same const
     // value.
     saw_const_match_error: Cell<bool>,
 
+    // This tracks if we emitted some diagnostic for a given const value, so that
+    // we will not subsequently issue an irrelevant lint for the same const
+    // value.
+    saw_const_match_lint: Cell<bool>,
+
+    // For backcompat we need to keep allowing non-structurally-eq types behind references.
+    // See also all the `cant-hide-behind` tests.
+    behind_reference: Cell<bool>,
+
     // inference context used for checking `T: Structural` bounds.
     infcx: InferCtxt<'a, 'tcx>,
 
     include_lint_checks: bool,
 }
 
+#[derive(Debug)]
+struct FallbackToConstRef;
```

Comment:
Perhaps document what returning `Err(FallbackToConstRef)` implies (and, importantly, that this is only meant to be used when `behind_reference.get() == true`, that part wasn't obvious).